File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/View/ApplyUserList/PanoApplyUserListView.m
//
//  PanoApplyUserListView.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/19.
//

#import "PanoApplyUserListView.h"
#import "PanoApplyUserListCell.h"
#import "PanoDefine.h"
#import "PanoMicInfo.h"
#import "PanoApplyAlertView.h"

#define headerHeight (isIphoneX() ? 44 : 20)

@interface PanoApplyUserListView () <UITableViewDelegate, UITableViewDataSource>

@end

@implementation PanoApplyUserListView
{
    UIView *coverView;
    UITableView *tableView;
    PanoApplyAlertView *footerView;
    UIView *headerView;
    BOOL isShowListView;
}
   
    
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initViews];
        [self initConstraints];
    }
    return self;
}

- (void)initViews {
    coverView = [UIView new];;
    self.userInteractionEnabled = true;
    self.frame = CGRectMake(0.0, 0, UIScreen.mainScreen.bounds.size.width, 0.0);
    footerView = [[PanoApplyAlertView alloc] initWithFrame:CGRectMake(0.0, 0, UIScreen.mainScreen.bounds.size.width, 35.0)];
    [footerView->countBtn addTarget:self action:@selector(showListView) forControlEvents:UIControlEventTouchUpInside];
    [self addSubview:footerView];
    tableView = [[UITableView alloc] init];
    [tableView registerClass:[PanoApplyUserListCell class] forCellReuseIdentifier:@"cellID"];
    tableView.tableFooterView = [UIView new];
    tableView.showsVerticalScrollIndicator = false;
    tableView.showsHorizontalScrollIndicator = false;
    tableView.delegate = self;
    tableView.dataSource = self;
    tableView.rowHeight = 60;
    tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    tableView.allowsSelection = false;
    [self addSubview:tableView];
    self.backgroundColor = [UIColor whiteColor];
}
    
- (void)initConstraints {
    [tableView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.top.left.right.mas_equalTo(0).insets(UIEdgeInsetsMake(headerHeight, 0, 0, 0));
    }];
    
    [footerView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.bottom.right.mas_equalTo(0);
        make.height.mas_equalTo(35);
        make.top.mas_equalTo(tableView.mas_bottom);
    }];
}


#pragma mark -- Extension
- (void)showListView {
    if (isShowListView) {
        [self dismissListView];
        return;
    }
    isShowListView = true;
    [self reloadView];
}

- (void)dismissListView {
    if (!isShowListView) {
        return;
    }
    isShowListView = false;
    [self dismiss];
}

- (void)reloadView {
    NSUInteger maxCount  = dataSource.count >= 4 ? 4 : dataSource.count;
    CGFloat height = headerHeight + 60 * maxCount + 35.0;
    self.frame = CGRectMake(0.0, 0, UIScreen.mainScreen.bounds.size.width, height);
    [tableView reloadData];
}

- (void)showInView:(UIView *)parentView {
    if (dataSource.count == 0) {
        [self dismiss];
        return;
    }
    if (![parentView.subviews containsObject:self]) {
        coverView.frame = parentView.bounds;
        coverView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.3];
        [coverView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissListView)]];
        [parentView addSubview:coverView];
        coverView.alpha = 0.01;
        [parentView addSubview:self];
    }
    
    self.alpha = 0.1;
    self.frame = CGRectMake(0, -PanoAppHeight, PanoAppWidth,  PanoAppHeight);
    [UIView animateWithDuration:0.4 animations:^{
        self.alpha = 1;
        self->coverView.alpha = 1;
        self.frame = CGRectMake(0, 0, PanoAppWidth,  PanoAppHeight);
    }];
    
    footerView->countLabel.text = [NSString stringWithFormat:@"%zd",dataSource.count];
    [self showListView];
}

- (void)dismiss {
    if (self.superview == nil) {
        return;
    }
    [UIView animateWithDuration:0.25 animations:^{
        self.alpha = 0.1;
        self->coverView.alpha = 0.1;
        self.frame = CGRectMake(0, -PanoAppHeight, PanoAppWidth,  0);
    } completion:^(BOOL finished) {
        [self->coverView removeFromSuperview];
        [self removeFromSuperview];
    }];
}

#pragma mark -- UITableViewDelegate, UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section  {
    return dataSource.count;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    PanoApplyUserListCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cellID" forIndexPath:indexPath];
    PanoMicInfo * micInfo = [self cellForUserInfo:indexPath];
    cell->nameLabel.text = [NSString stringWithFormat:@"'%@' %@", micInfo.user.userName ? micInfo.user.userName : NSLocalizedString(@"未知", nil), NSLocalizedString(@"正在申请上麦", nil)];
    cell->acceptBlock = ^{
        [self dismissListView];
        [self.delegate onClickAcceptButton:micInfo];
    };
    cell->rejectBlock = ^{
        [self dismissListView];
        [self.delegate onClickRejectButton:micInfo];
    };
    return cell;
}

- (PanoMicInfo *)cellForUserInfo:(NSIndexPath *)indexPath {
    if (indexPath.row < dataSource.count) {
        return dataSource[indexPath.row];
    }
    return nil;
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/View/ApplyUserList/PanoApplyUserListView.swift
//
//  PanoApplyUserListView.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/11/3.
//

import UIKit

protocol PanoApplyUserListDelegate : NSObjectProtocol {
    
    func onClickAcceptButton(info: PanoMicInfo!);
    
    func onClickRejectButton(info: PanoMicInfo!);
    
}


class PanoApplyUserListView: UIControl {

    var coverView: UIView!
    var tableView: UITableView!
    var footerView: PanoApplyAlertView!
    var dataSource: [PanoMicInfo]!
    let headerHeight: CGFloat = PanoIsIphoneX() ? 44 : 20;
    var headerView: UIView!
    var isShowListView: Bool = false
    weak var delegate: PanoApplyUserListDelegate!
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.initViews()
        self.initConstraints()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func initViews() {
        coverView = UIView();
        self.isUserInteractionEnabled = true
        self.frame = CGRect(x: 0.0, y: 0, width: PanoAppWidth, height: 0.0)
        footerView = PanoApplyAlertView(frame: CGRect(x: 0, y: 0, width: PanoAppWidth, height: 35));
        footerView.countBtn.addTarget(self, action: #selector(showListView), for: .touchUpInside)
//        footerView.backgroundColor = UIColor.clear
        self.addSubview(footerView)
        tableView = UITableView()
        tableView.register(PanoApplyUserListCell.self, forCellReuseIdentifier: "cellID")
        tableView.tableFooterView = UIView()
        tableView.showsVerticalScrollIndicator = false
        tableView.showsHorizontalScrollIndicator = false
        tableView.delegate = self
        tableView.dataSource = self
        tableView.rowHeight = 60
        tableView.separatorStyle = .none
        tableView.allowsSelection = false
        self.addSubview(tableView)
        self.backgroundColor = UIColor.white
    }
    
    func initConstraints() {
        tableView.snp.remakeConstraints { (make) in
            make.top.left.right.equalToSuperview().inset(UIEdgeInsets(top: headerHeight, left: 0, bottom: 0, right: 0))
        }
        footerView.snp.remakeConstraints { (make) in
            make.left.bottom.right.equalToSuperview();
            make.height.equalTo(35)
            make.top.equalTo(tableView.snp.bottom)
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        self.roundCorners(corners: [.bottomLeft, .bottomRight], radius: 15)
    }
}

extension PanoApplyUserListView {
    
    @objc private func showListView() {
        if isShowListView {
            dismissListView()
            return
        }
        isShowListView = true
        reloadView()
    }
    
    @objc private func dismissListView()  {
        if !isShowListView {
            return
        }
        isShowListView = false
        self.dismiss()
    }
    
    func reloadView() {
        let maxCount  = dataSource.count >= 4 ? 4 : dataSource.count
        let height = headerHeight + 60 * CGFloat(maxCount) + 35.0;
        self.frame = CGRect(x: 0.0, y: 0, width: PanoAppWidth, height: height)
        self.tableView.reloadData()
    }
    
    func showInView(parentView: UIView) {
        if dataSource.count == 0 {
            dismiss()
            return
        }
        if !(parentView.subviews.contains(self)) {
            coverView.frame = parentView.bounds
            coverView.backgroundColor = UIColor.black.withAlphaComponent(0.3)
            coverView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(dismissListView)))
            parentView.addSubview(coverView)
            coverView.alpha = 0.01;
            parentView.addSubview(self)
        }
        
        self.alpha = 0.1;
        self.frame = CGRect(x: 0, y: -PanoAppHeight, width: PanoAppWidth, height: PanoAppHeight)
        UIView.animate(withDuration: 0.4) {
            self.alpha = 1
            self.coverView.alpha = 1
            self.frame = CGRect(x: 0, y: 0, width: PanoAppWidth, height: PanoAppHeight)
        }
        
        footerView.countLabel.text = "\(dataSource.count)"
        showListView()
    }
    
    func dismiss() {
        if self.superview == nil {
            return
        }
        UIView.animate(withDuration: 0.3) {
            self.alpha = 1
            self.coverView.alpha = 1
            self.frame = CGRect(x: 0, y: -PanoAppHeight, width: PanoAppWidth, height: 0)
        } completion: { (flag) in
            self.coverView.removeFromSuperview()
            self.removeFromSuperview()
        }
    }
}

extension PanoApplyUserListView: UITableViewDelegate, UITableViewDataSource{
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataSource.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellID", for: indexPath) as! PanoApplyUserListCell
        let micInfo = cellForUserInfo(indexPath: indexPath);
        cell.nameLabel.text = "'" + (micInfo?.user?.userName ??  NSLocalizedString("未知", comment: "")) + "'" + NSLocalizedString("正在申请上麦", comment: "")
        cell.acceptBlock = {
            self.dismissListView()
            self.delegate.onClickAcceptButton(info: micInfo);
        }
        cell.rejectBlock = {
            self.dismissListView()
            self.delegate.onClickRejectButton(info: micInfo);
        }
        return cell
    }
    
    func cellForUserInfo(indexPath: IndexPath) -> PanoMicInfo? {
        let micInfos = dataSource
        if indexPath.row < micInfos?.count ?? 0 {
            let info = micInfos?[indexPath.row]
            return info
        }
        return nil
    }
}
